File: agent/enforcer/firewall.go
package enforcer

import (
	"fmt"
	"log"
	"os/exec"
	"strconv"
	"sync"
)

// FirewallEnforcer manages nftables rules for port protection
type FirewallEnforcer struct {
	tunName        string
	tableName      string // nftables table name
	chainName      string // nftables chain name
	protectedPorts map[int]*ProtectionRule
	mu             sync.RWMutex
}

// ProtectionRule tracks a protected port
type ProtectionRule struct {
	Port     int
	Protocol string
	Active   bool
	RuleIDs  []string // Track which nftables rules were added
}

// NewFirewallEnforcer creates a new firewall enforcer
func NewFirewallEnforcer(tunName string) *FirewallEnforcer {
	fe := &FirewallEnforcer{
		tunName:        tunName,
		tableName:      "ztna",
		chainName:      "input_filter",
		protectedPorts: make(map[int]*ProtectionRule),
	}

	// Initialize nftables table and chain
	if err := fe.initializeNftables(); err != nil {
		log.Printf("⚠️  Failed to initialize nftables: %v", err)
	}

	return fe
}

// initializeNftables sets up the nftables table and chain
func (f *FirewallEnforcer) initializeNftables() error {
	log.Println("Initializing nftables table and chain...")

	// Check if nftables is available
	if err := exec.Command("nft", "--version").Run(); err != nil {
		return fmt.Errorf("nftables not available: %v (install with: sudo apt install nftables)", err)
	}

	// Create table if it doesn't exist
	// Using 'inet' family for both IPv4 and IPv6
	cmd := exec.Command("nft", "add", "table", "inet", f.tableName)
	if err := cmd.Run(); err != nil {
		// Table might already exist, that's OK
		log.Printf("Table creation: %v (may already exist)", err)
	}

	// Create chain with high priority to hook into input
	// Priority -150 is before conntrack (priority -200) but after other filters
	cmd = exec.Command("nft", "add", "chain", "inet", f.tableName, f.chainName,
		"{ type filter hook input priority -150 ; policy accept ; }")
	if err := cmd.Run(); err != nil {
		log.Printf("Chain creation: %v (may already exist)", err)
	}

	log.Printf("✓ nftables initialized: table=%s, chain=%s", f.tableName, f.chainName)
	return nil
}

// ProtectPort applies nftables rules to protect a port
func (f *FirewallEnforcer) ProtectPort(port int, protocol string) error {
	f.mu.Lock()
	defer f.mu.Unlock()

	// Check if already protected
	if rule, exists := f.protectedPorts[port]; exists && rule.Active {
		return fmt.Errorf("port %d is already protected", port)
	}

	portStr := strconv.Itoa(port)

	log.Printf("Applying nftables rules for port %d (%s)...", port, protocol)

	// nftables handles all rules in a single chain with better performance
	// We use a named handle system for easier rule management

	// Rule 1: ACCEPT from localhost (lo interface)
	handle1, err := f.addRule(fmt.Sprintf(
		"iifname \"lo\" tcp dport %s accept",
		portStr,
	))
	if err != nil {
		return fmt.Errorf("failed to add localhost rule: %v", err)
	}

	// Rule 2: ACCEPT from TUN interface
	handle2, err := f.addRule(fmt.Sprintf(
		"iifname \"%s\" tcp dport %s accept",
		f.tunName, portStr,
	))
	if err != nil {
		// Rollback rule 1
		f.deleteRuleByHandle(handle1)
		return fmt.Errorf("failed to add TUN rule: %v", err)
	}

	// Rule 3: DROP everything else to this port
	handle3, err := f.addRule(fmt.Sprintf(
		"tcp dport %s drop",
		portStr,
	))
	if err != nil {
		// Rollback
		f.deleteRuleByHandle(handle1)
		f.deleteRuleByHandle(handle2)
		return fmt.Errorf("failed to add DROP rule: %v", err)
	}

	// Store the rule with handles
	f.protectedPorts[port] = &ProtectionRule{
		Port:     port,
		Protocol: protocol,
		Active:   true,
		RuleIDs:  []string{handle1, handle2, handle3},
	}

	log.Printf("✓ Port %d is now protected (handles: %s, %s, %s)", port, handle1, handle2, handle3)
	return nil
}

// UnprotectPort removes nftables rules for a port
func (f *FirewallEnforcer) UnprotectPort(port int) error {
	f.mu.Lock()
	defer f.mu.Unlock()

	rule, exists := f.protectedPorts[port]
	if !exists || !rule.Active {
		return fmt.Errorf("port %d is not protected", port)
	}

	log.Printf("Removing nftables rules for port %d...", port)

	errors := []error{}

	// Delete rules by their handles
	for _, handle := range rule.RuleIDs {
		if err := f.deleteRuleByHandle(handle); err != nil {
			errors = append(errors, err)
		}
	}

	if len(errors) > 0 {
		log.Printf("⚠️  Some rules failed to remove: %v", errors)
	}

	// Mark as inactive
	rule.Active = false
	delete(f.protectedPorts, port)

	log.Printf("✓ Port %d is now unprotected", port)
	return nil
}

// GetProtectedPorts returns list of currently protected ports
func (f *FirewallEnforcer) GetProtectedPorts() []int {
	f.mu.RLock()
	defer f.mu.RUnlock()

	ports := make([]int, 0, len(f.protectedPorts))
	for port, rule := range f.protectedPorts {
		if rule.Active {
			ports = append(ports, port)
		}
	}
	return ports
}

// IsPortProtected checks if a port is protected
func (f *FirewallEnforcer) IsPortProtected(port int) bool {
	f.mu.RLock()
	defer f.mu.RUnlock()

	rule, exists := f.protectedPorts[port]
	return exists && rule.Active
}

// addRule adds an nftables rule and returns its handle
func (f *FirewallEnforcer) addRule(ruleSpec string) (string, error) {
	// Add rule with echo option to get handle
	cmd := exec.Command("nft", "-ae", "add", "rule", "inet", f.tableName, f.chainName, ruleSpec)
	output, err := cmd.CombinedOutput()
	if err != nil {
		return "", fmt.Errorf("nft error: %v, output: %s", err, string(output))
	}

	// Output format: "add rule inet ztna input_filter tcp dport 5173 accept # handle 5"
	// Parse to extract handle number
	outputStr := string(output)
	if idx := lastIndex(outputStr, "handle "); idx != -1 {
		handleStr := outputStr[idx+7:] // Skip "handle "
		// Trim any whitespace/newlines
		handleStr = trimSpace(handleStr)
		return handleStr, nil
	}

	// If we can't parse handle, return empty (will use alternative deletion method)
	return "", nil
}

// deleteRuleByHandle deletes an nftables rule by handle
func (f *FirewallEnforcer) deleteRuleByHandle(handle string) error {
	if handle == "" {
		// No handle available, skip
		return nil
	}

	cmd := exec.Command("nft", "delete", "rule", "inet", f.tableName, f.chainName, "handle", handle)
	output, err := cmd.CombinedOutput()
	if err != nil {
		return fmt.Errorf("nft error: %v, output: %s", err, string(output))
	}
	return nil
}

// Helper functions
func lastIndex(s, substr string) int {
	for i := len(s) - len(substr); i >= 0; i-- {
		if s[i:i+len(substr)] == substr {
			return i
		}
	}
	return -1
}

func trimSpace(s string) string {
	// Simple whitespace trimmer
	start := 0
	end := len(s)

	for start < end && (s[start] == ' ' || s[start] == '\n' || s[start] == '\t' || s[start] == '\r') {
		start++
	}

	for end > start && (s[end-1] == ' ' || s[end-1] == '\n' || s[end-1] == '\t' || s[end-1] == '\r') {
		end--
	}

	return s[start:end]
}

func contains(haystack, needle string) bool {
	for i := 0; i <= len(haystack)-len(needle); i++ {
		if haystack[i:i+len(needle)] == needle {
			return true
		}
	}
	return false
}

func (f *FirewallEnforcer) CleanupAll() {
	f.mu.Lock()
	defer f.mu.Unlock()

	log.Printf("🧹 Cleaning up all firewall rules in table %s...", f.tableName)

	// Deleting the table removes all chains and rules within it
	cmd := exec.Command("nft", "delete", "table", "inet", f.tableName)
	if err := cmd.Run(); err != nil {
		log.Printf("⚠️  Note: Could not delete table (it may have been already removed): %v", err)
	} else {
		log.Printf("✓ nftables table %s removed", f.tableName)
	}

	// Reset the local map
	f.protectedPorts = make(map[int]*ProtectionRule)
}

File: services/agent/src/firewall.rs
use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use tokio::process::Command;
use tokio::sync::Mutex;
use tracing::{info, warn};

#[derive(Deserialize)]
pub struct FirewallPolicy {
    pub action: String,
    pub protected_ports: Vec<PortRule>,
}

#[derive(Deserialize, Serialize, Clone, PartialEq, Eq, Hash, Debug)]
pub struct PortRule {
    pub port: u16,
    pub protocol: String,
}

#[derive(Serialize, Deserialize, Default, Debug)]
pub struct FirewallState {
    pub protected_ports: Vec<PortRule>,
}

struct ProtectionEntry {
    port: u16,
    protocol: String,
    rule_handles: Vec<u64>,
}

pub struct FirewallEnforcer {
    tun_name: String,
    table_name: String,
    chain_name: String,
    protected: Mutex<HashMap<u16, ProtectionEntry>>,
}

impl FirewallEnforcer {
    pub fn new(tun_name: &str) -> Self {
        Self {
            tun_name: tun_name.to_string(),
            table_name: "ztna".to_string(),
            chain_name: "input_filter".to_string(),
            protected: Mutex::new(HashMap::new()),
        }
    }

    /// Create the nftables table and chain if they don't exist.
    pub async fn initialize(&self) -> Result<()> {
        // Create table
        run_nft_allow_exists(&format!(
            "add table inet {}",
            self.table_name
        ))
        .await
        .context("failed to create nftables table")?;

        // Create input chain with filter hook
        run_nft_allow_exists(&format!(
            "add chain inet {} {} {{ type filter hook input priority 0 ; policy accept ; }}",
            self.table_name, self.chain_name
        ))
        .await
        .context("failed to create nftables chain")?;

        info!(
            "nftables initialized: table={} chain={}",
            self.table_name, self.chain_name
        );
        Ok(())
    }

    /// Diff current protected ports vs desired and add/remove as needed.
    pub async fn sync_policy(&self, desired: &[PortRule]) -> Result<()> {
        let mut protected = self.protected.lock().await;

        let desired_ports: HashMap<u16, &PortRule> =
            desired.iter().map(|r| (r.port, r)).collect();

        // Remove ports no longer desired
        let stale: Vec<u16> = protected
            .keys()
            .filter(|p| !desired_ports.contains_key(p))
            .copied()
            .collect();

        for port in stale {
            if let Some(entry) = protected.remove(&port) {
                if let Err(e) = self.delete_rules(&entry.rule_handles).await {
                    warn!("failed to unprotect port {}: {}", port, e);
                }
                info!("unprotected port {}/{}", port, entry.protocol);
            }
        }

        // Add new ports
        for rule in desired {
            if protected.contains_key(&rule.port) {
                continue;
            }
            match self.protect_port(rule.port, &rule.protocol).await {
                Ok(handles) => {
                    protected.insert(
                        rule.port,
                        ProtectionEntry {
                            port: rule.port,
                            protocol: rule.protocol.clone(),
                            rule_handles: handles,
                        },
                    );
                    info!("protected port {}/{}", rule.port, rule.protocol);
                }
                Err(e) => {
                    warn!("failed to protect port {}/{}: {}", rule.port, rule.protocol, e);
                }
            }
        }

        Ok(())
    }

    /// Add three nftables rules for a port:
    /// 1. Accept from loopback
    /// 2. Accept from TUN interface
    /// 3. Drop everything else
    async fn protect_port(&self, port: u16, protocol: &str) -> Result<Vec<u64>> {
        let proto = protocol.to_lowercase();
        let mut handles = Vec::new();

        // Rule 1: accept from lo
        let h1 = self
            .add_rule(&format!(
                "iifname \"lo\" {} dport {} accept",
                proto, port
            ))
            .await
            .context("lo accept rule")?;
        handles.push(h1);

        // Rule 2: accept from TUN
        let h2 = match self
            .add_rule(&format!(
                "iifname \"{}\" {} dport {} accept",
                self.tun_name, proto, port
            ))
            .await
        {
            Ok(h) => h,
            Err(e) => {
                // Rollback rule 1
                let _ = self.delete_rules(&handles).await;
                return Err(e).context("tun accept rule");
            }
        };
        handles.push(h2);

        // Rule 3: drop all other traffic to this port
        let h3 = match self
            .add_rule(&format!("{} dport {} drop", proto, port))
            .await
        {
            Ok(h) => h,
            Err(e) => {
                let _ = self.delete_rules(&handles).await;
                return Err(e).context("drop rule");
            }
        };
        handles.push(h3);

        Ok(handles)
    }

    /// Add a single rule and return its handle.
    async fn add_rule(&self, rule_expr: &str) -> Result<u64> {
        let cmd = format!(
            "-ae add rule inet {} {} {}",
            self.table_name, self.chain_name, rule_expr
        );
        let output = run_nft_output(&cmd).await?;

        // nft -ae outputs "# handle <N>" — extract the handle number
        parse_handle(&output)
            .ok_or_else(|| anyhow::anyhow!("could not parse rule handle from nft output"))
    }

    /// Delete rules by their handles.
    async fn delete_rules(&self, handles: &[u64]) -> Result<()> {
        for handle in handles {
            let cmd = format!(
                "delete rule inet {} {} handle {}",
                self.table_name, self.chain_name, handle
            );
            if let Err(e) = run_nft(&cmd).await {
                warn!("failed to delete rule handle {}: {}", handle, e);
            }
        }
        Ok(())
    }

    /// Remove protection for a specific port (public API for on-demand removal).
    #[allow(dead_code)]
    pub async fn unprotect_port(&self, port: u16) -> Result<()> {
        let mut protected = self.protected.lock().await;
        if let Some(entry) = protected.remove(&port) {
            self.delete_rules(&entry.rule_handles).await?;
            info!("unprotected port {}/{}", port, entry.protocol);
        }
        Ok(())
    }

    /// Return current state for persistence.
    pub async fn get_state(&self) -> FirewallState {
        let protected = self.protected.lock().await;
        FirewallState {
            protected_ports: protected
                .values()
                .map(|e| PortRule {
                    port: e.port,
                    protocol: e.protocol.clone(),
                })
                .collect(),
        }
    }

    /// Restore firewall rules from persisted state (called on startup).
    pub async fn restore_from_state(&self, state: &FirewallState) -> Result<()> {
        if state.protected_ports.is_empty() {
            return Ok(());
        }
        info!(
            "restoring firewall state for {} ports",
            state.protected_ports.len()
        );
        self.sync_policy(&state.protected_ports).await
    }

    /// Delete the entire nftables table (cleanup on shutdown).
    pub async fn cleanup_all(&self) {
        let cmd = format!("delete table inet {}", self.table_name);
        match run_nft(&cmd).await {
            Ok(()) => info!("cleaned up nftables table {}", self.table_name),
            Err(e) => warn!("failed to cleanup nftables table: {}", e),
        }
    }
}

/// Handle an inbound `firewall_policy` message from the connector.
pub async fn handle_firewall_policy(
    payload: &[u8],
    enforcer: &FirewallEnforcer,
) -> Result<()> {
    let policy: FirewallPolicy = serde_json::from_slice(payload)
        .context("failed to parse firewall_policy message")?;

    match policy.action.as_str() {
        "sync" => {
            enforcer.sync_policy(&policy.protected_ports).await?;
            // Persist state after sync
            let state = enforcer.get_state().await;
            if let Err(e) = crate::persistence::save_firewall_state(&state) {
                warn!("failed to persist firewall state: {}", e);
            }
        }
        other => {
            warn!("unknown firewall policy action: {}", other);
        }
    }

    Ok(())
}

/// Run an nft command, returning an error if it fails.
async fn run_nft(args: &str) -> Result<()> {
    let output = Command::new("nft")
        .args(args.split_whitespace())
        .output()
        .await
        .context("failed to execute nft")?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        anyhow::bail!("nft {} failed: {}", args, stderr.trim());
    }
    Ok(())
}

/// Run an nft command and ignore "File exists" errors.
async fn run_nft_allow_exists(args: &str) -> Result<()> {
    match run_nft(args).await {
        Ok(()) => Ok(()),
        Err(e) => {
            let msg = e.to_string();
            if msg.contains("File exists") {
                Ok(())
            } else {
                Err(e)
            }
        }
    }
}

/// Run an nft command and return stdout.
async fn run_nft_output(args: &str) -> Result<String> {
    let output = Command::new("nft")
        .args(args.split_whitespace())
        .output()
        .await
        .context("failed to execute nft")?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        anyhow::bail!("nft {} failed: {}", args, stderr.trim());
    }
    Ok(String::from_utf8_lossy(&output.stdout).to_string())
}

/// Parse "# handle <N>" from nft -ae output.
fn parse_handle(output: &str) -> Option<u64> {
    for line in output.lines() {
        if let Some(idx) = line.find("# handle ") {
            let rest = &line[idx + "# handle ".len()..];
            if let Ok(h) = rest.trim().parse::<u64>() {
                return Some(h);
            }
        }
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_handle() {
        assert_eq!(parse_handle("# handle 42"), Some(42));
        assert_eq!(parse_handle("  # handle 123  "), Some(123));
        assert_eq!(parse_handle("no handle here"), None);
        assert_eq!(
            parse_handle("add rule inet ztna input_filter ...\n# handle 7"),
            Some(7)
        );
        assert_eq!(
            parse_handle("add rule inet ztna input_filter iifname \"lo\" tcp dport 22 accept # handle 5"),
            Some(5)
        );
    }
}
